feat!: Upgrade `Kraftfile` specification to v0.6

nderjung opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

  Kraftkit follows the same guidelines as the Unikraft Open Source Project.

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

This PR introduces a new `Kraftfile` specification which makes new adjustments to the user's workflow such that it is easier to run applications in binary compatibility mode and with a template.  At a high-level, the following adjustments are made to the specification:

- A new top-level `runtime` element to be used for defining the user of an elfloader or existing pre-built unikernel image.  Similar to other component-type elements like `unikraft` and `libraries`, the `runtime` element is intended to allow for a short-hand and long-hand syntax where a `source`, `version` and set of `kconfig` options can be used to qualify the elfloader application.
- Abbreviations of the `specification`, `architecture` and `platform` element names such that can be used in short-hand syntax as `spec`, `arch` and `plat`, respectively.
- Top-level `volumes` element which is used for defining user-specified paths which are to be used during runtime of a unikernel.
- Top-level `rootfs` element which is used for defining the root filesystem of the user.  The provided implementation allows a user to serialize a filesystem from three new contexts: an existing CPIO archive (passthrough), a path to a directory, and a `Dockerfile` which is constructed via [BuildKit](https://github.com/moby/buildkit).
- Top-level `cmd` which is a string or list which represents the command-line argument to execute for the unikernel as a way to porclain the Kraftfile and the intended application usecase.

Aside from these introductions to the v0.6 specification, the underlying parser is adjusted to handle reading both the specifications and defaulting to v0.6 as the latest.  These adjustments are subsequently implemented through the PR.


### High-level `Kraftfile` v0.6 specification changes

#### A. Introduction of short-hand syntaxes

1. In addition to bumping the specification version to v0.6, the `specification` element can be abbreviated to `spec`:

   ```yaml
   spec: v0.6
   ```

2. Within the list of `targets`, the `architecture` and `platform` attributes can be abbreviated to `arch` and `plat`, respectively, and be used interchangeably:

   ```yaml
   targets:
   - plat: qemu
     arch: x86_64
   ```

   ```yaml
   targets:
   - plat: qemu
     architecture: x86_64
   ```

   ```yaml
   targets:
   - platform: qemu
     arch: x86_64
   ```

3. The list of targets can now accept an even shorter syntax where only the architecture and platform are desired in the list:

   ```yaml
   targets:
   - qemu/x86_64
   ```

   This shorthand syntax can be mixed with full target elements:

   ```yaml
   targets:
   - qemu/x86_64
   - platform: qemu
     architecture: arm64
   - plat: fc
     arch: x86_64
   ```

#### C. Introduction of a `cmd` element for specifying program operation

The `cmd` is array or string element which can be used for setting default arguments to be used during the instantiation of a new unikernel instance:

1. Specified as an in-line array:
   ```yaml
   spec: v0.6

   cmd: ["-c", "/nginx/conf/nginx.conf"]
   ```

2. Specified as a multi-line array:
   ```yaml
   spec: v0.6

   cmd:
   - -c
   - /nginx/conf/nginx.conf
   ```

3. Specified as a string:
   ```yaml
   spec: v0.6

   cmd: "-c /nginx/conf/nginx.conf"
   ```

#### C. Introduction to `runtime` element for accessing pre-built images

The `runtime` element is a new top-level element which has been designed to ease the declaration of a previously constructed unikernel image in the project.  In many cases, specifically instances where the user is writing in a high-level language such as Python or where they wish to use a pre-built Linux userpace image, the user does not care for wanting to build a unikernel image from scratch.  Instead, the goal is to leverage pre-built unikernel images which are distributed using the [OCI Package Manager](https://github.com/unikraft/kraftkit/pull/339). 

Similar to the top-level `unikraft` and `template` elements (see below), the `runtime` element can be specified using a short-hand syntax or a long-hand syntax for greater control or preference.

1. In short-hand syntax, simply referencing the pre-built unikernel image is possible by declaring:
   ```yaml
   spec: v0.6
   
   runtime: unikraft.org/nginx:latest
   ```
   In the above example, the `runtime` element refers to an OCI archive for NGINX.  In a `kraft run` scenario, this is simply pulled and executed verbatim.  Paired with the ability to customize the `rootfs` element or `volumes` elemen, the `runtime` element becomes more pragmatic:
   ```yaml
   spec: v0.6
   
   runtime: unikraft.org/nginx:latest
   
   rootfs: ./
   ```
   Or with a volume, which simply "overwrites" an existing directory already located within NGINX's root filesystem that is distributed as part of the OCI image:
   ```yaml
   spec: v0.6
   
   runtime: unikraft.org/nginx:latest
   
   volumes:
   - ./html:/nginx/html
   ```
   In the above snippet, the user customizes only the directory which serves HTML with a path `./html` which is located on their host.
   
   Note that this usecase is not limited to a single application and applies to the ELF Loader application and customizing the loaded application which is part of a root filesystem:
   ```yaml
   spec: v0.6
   
   # Default ELF Loader application
   runtime: loaders.unikraft.org/default:latest 
   
   # Path must contain the Linux ELF binary, e.g. at ./rootfs/helloworld
   rootfs: ./rootfs
   
   cmd: ["/helloworld"]
   ```
   or a high-level programming language such as python3:
   ```yaml
   spec: v0.6
   
   runtime: unikraft.org/python3:latest
   
   rootfs: ./Dockerfile
   
   cmd: ["/main.py"]
   ```
   In the above snippet, the user has created a `Dockerfile` which contains the necessary root filesystem and included their application at `/main.py`.

2. The long-hand syntax allows specifically setting `source`, `version` and `kconfig` options which are requirements of the runtime:
   ```yaml
   spec: v0.6
   
   runtime:
     source: unikraft.org/nginx
     version: latest
     kconfig:
       CONFIG_LIB9PFS: 'y'
    ```
    In the above example, the a pre-built NGINX unikernel has been requested as the runtime with the explicit option of having the 9P Filesystem enabled.

#### D. Fully realizing the `template` element

There were several issues related to the `template` element, including the fact that it was not properly seeded, or transformed, from the `Kraftfile` into the application structure. This has now been tested end-to-end and enables:

1. Workflows where a user may wish to reference an existing application repository and make customizations, for example, with the [ELF Loader](https://github.com/unikraft/app-elfloader.git) application, making changes to the core configuration and adding additional targets:
   ```yaml
   spec: v0.6
   
   template: https://github.com/unikraft/app-elfloader.git
   
   unikraft:
     version: stable
     kconfig:
       CONFIG_LIBUKLIBPARAM: 'y'
       CONFIG_LIBNOLIBC: 'y'
       CONFIG_LIBNOLIBC_UKDEBUG_ASSERT: 'y'
       CONFIG_LIBPOSIX_EVENT: 'y'
       CONFIG_LIBPOSIX_FUTEX: 'y'
       CONFIG_LIBPOSIX_MMAP: 'y'
       CONFIG_LIBPOSIX_PROCESS: 'y'
       CONFIG_LIBPOSIX_PROCESS_PIDS: 'y'
       CONFIG_LIBPOSIX_PROCESS_MAX_PID: 31
       CONFIG_LIBPOSIX_PROCESS_CLONE: 'y'
       CONFIG_LIBPOSIX_SOCKET: 'y'
       CONFIG_LIBPOSIX_SYSINFO: 'y'
       CONFIG_LIBPOSIX_TIME: 'y'
       CONFIG_LIBPOSIX_USER: 'y'
       CONFIG_LIBUKSIGNAL: 'y'
       CONFIG_LIBSYSCALL_SHIM: 'y'
       CONFIG_LIBVFSCORE_AUTOMOUNT_ROOTFS: 'y'
       CONFIG_LIBVFSCORE_ROOTFS_INITRD: 'y'
       CONFIG_LIBVFSCORE_ROOTFS: "initrd"
       CONFIG_APPELFLOADER_VFSEXEC_EXECBIT: 'n'

   targets:
   - plat: qemu
     arch: x86_64
     kconfig:
       CONFIG_VIRTIO_BUS: 'n'
   - plat: firecracker
     arch: x86_64
   ```

3. Similar to other component elements (e.g. `unikraft` and `libraries`), the `template` element can be expressed in a longer format:
   ```yaml
   spec: v0.6
   
   template:
     source: https://github.com/unikraft/app-elfloader.git
     version: stable
   ```

### New workflow scenarios

The following workflow scenarios were used during testing.

#### A. Using a pre-built unikernel and customizing the root filesystem

Customizing the whole root filesystem is now possible by supplying either:

1. A path representing a prebuilt CPIO archive file:
   ```yaml
   spec: v0.6

   runtime: unikraft.org/nginx:latest

   rootfs: ./initramfs.cpio
   ```

4. A directory which will be dynamically serialized into a CPIO archive:
   ```yaml
   spec: v0.6

   runtime: unikraft.org/nginx:latest

   rootfs: ./rootfs/
   ```

5. Or, a `Dockerfile` which will be built via BuildKit and then serialized into a new CPIO archive:
   ```yaml
   spec: v0.6

   runtime: unikraft.org/nginx:latest

   rootfs: ./Dockerfile
   ```
   To use BuildKit, set the address within the new `buildkit_host` user setting, the new environmental variable `KRAFTKIT_BUILDKIT_HOST` or the command-line flag `--buildkit-host`.  This defaults to `unix:///run/buildkit/buildkitd.sock`.


#### B. Using a pre-built unikernel and mounting a path within the root filesystem

1. Specifying via short-hand syntax a map between a directory on the host and a path within the instance:
   ```yaml
   spec: v0.6

   runtime: unikraft.org/nginx:latest

   volumes:
   - ./src:/nginx/html
   ```

2. Using a more verbose (long-hand) syntax and mapping between the mount on the host and a path within the instance:
   ```yaml
   spec: v0.6

   runtime: unikraft.org/nginx:latest

   volumes:
   - source: ./src
     destination: /nginx/html
   ```

nderjung committed:
refactor(unikraft)!: Deprecate `lib.Libraries`

This commit removes the type alias for `lib.Libraries` and expands it to
its proper and full type.  This reduces ambiguity in its use with the
similar type `lib.Library` which is an interface.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(app)!: Convert interface to struct pointer

Simplify the use of the attribute, removing unnecessary type-casting.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(app)!: Convert interface to struct pointer

Simplify the use of the attribute, removing unnecessary type-casting.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(app): Fix nil-pointer exceptions

For `Kraftfile`s which have an uninitialized `unikraft` directive,
these line would otherwise throw a nil-pointer exception.  Fix by
simply checking if the attribute has been set.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(run): Increase runner specificity

This runner is used only when a project is built from source with access
to Unikraft' source code.  Ensure this by making necessary adjustments
to the `Runnable` method.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(run): Propagate workdir to top-level structure

This allows re-use without re-instantiation.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(schema): Introduce v0.6 Kraftfile specification

This commit introduces a new Kraftfile specification version, v0.6,
which includes the following major adjustments:

- A new top-level `runtime` element to be used for defining the user of
  an elfloader or existing pre-built unikernel image.  Similar to other
  component-type elements like `unikraft` and `libraries`, the `runtime`
  element is intended to allow for a short-hand and long-hand syntax
  where a `source`, `version` and set of `kconfig` options can be used
  to qualify the elfloader application.
- A new top-level `entrypoint` element which is used in conjunction with
  the `runtime` element which is used to specifty the path of the Linux
  userspace application to be run ontop of an elfloader application in
  binary compatibility mode.  Alternatively, the `entrypoint` element
  can be used as an alternative to specifying a source file representing
  an application, for example with Go- or Rust-based applications.
- Abbreviations of the `specification`, `architecture` and `platform`
  element names such that can be used in short-hand syntax as `spec`,
  `arch` and `plat`, respectively.
- Top-level `volumes` element which is used for defining user-specified
  paths which are to be used during runtime of a unikernel.
- Top-level `rootfs` element which is used for defining the root file-
  system of the user.  An ultimate implementation is able to serialize
  the provided string based on context, for example if the user
  specifies a directory or simply a path to an initramfs cpio archive.
- Top-level `cmd` which is a string or list which represents the
  command-line argument to execute for the unikernel as a way to
  porclain the Kraftfile and the intended application usecase.

Aside from these introductions to the v0.6 specification, the underlying
parser is adjusted to handle reading both the specifications and
defaulting to v0.6 as the latest.  These adjustments are not, however,
realised in the application component as this left to future work.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat: Populate application `cmd` from `Kraftfile`

This commit realises the adjustments to the v0.6 spec for the `cmd`
element which is intended to be used as part of the runtime of the
unikernel.  For now, this commit simply reads this information into
the application component.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat: Populate application `rootfs` from `Kraftfile`

This commit realises the adjustments to the v0.6 spec for the `rootfs`
element which is intended to be used as part of both the build- and
run-time of the unikernel.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat!: Introduce dynamic initramfs serialization for root filesystems

This commit introduces a refactor of the mechanics behind serializing an
input "rootfs" such that it is a usable initramfs.  This change starts
with the adjustment of the `initrd` package which is now based on an
interface `Initrd` which is used to introduce several underlying
implementations that can serialize a provided input.  The following
possible inputs are:

- A path to a directory which will act as the rootfs;
- A `Dockerfile` which can be built via buildkit; and,
- For compatibility of the interface, an existing file representing a
  CPIO archive.

The adjustment to this interface has been made across the project
including relevant tools.  Importantly, the GitHub Action has also been
updated to set up BuildKit as a "sidecar" via the composite type.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(cmdfactory): Introduce a `hidden` attribute to enable hiding flags

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat: Implement `runtime` element of the v0.6 spec update

This commit correctly handles the scenario where a user has set the
`runtime` attribute of the `Kraftfile` as part of the v0.6 specification
which allows the user to specify a pre-built unikernel image.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor: Use option-methods for ELFLoader instantiation

This commit refactors the instantiation of an ELF Loader application
based on option-method pattern.  Additionally, attributes `kernel`,
`name`, `source` and rootfs are provided.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(target): Introduce target aliasing

This commit introduces aliasing to list of targets such that
platform/architecture tuples can be defined simply as:

```yaml
targets:
- qemu/x86_64
```

For simpler projects with a single target, this makes the supplied
`Kraftfile` smaller and easier to read.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(target): Fix setting architecture and platform kconfig names

Previously, the KConfig option values for the architecture and platform
were not set unless the kconfig attribute was nil.  When the user
decides to create a target with specific KConfig options, these implict
values were not set, resulting in an undefined platform/architecture in
Unikraft's build.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(target): Handle a direct path to the kernel

When the value of the kernel attribute is set, discover its name by
reading the file name instead of using the pre-defined naming scheme.

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc commented on unikraft/app/application.go:
```diff
@@ -48,7 +48,7 @@ type Application interface {
 
 	// Libraries returns the application libraries' configurations
-	Libraries(ctx context.Context) (lib.Libraries, error)
+	Libraries(ctx context.Context) (map[string]*lib.LibraryConfig, error)
```
In the commit (commit 1) message for this change: `ambigiuity` -> `ambiguity`

craciunoiuc commented on oci/image.go:
```diff
@@ -205,6 +205,11 @@ func (image *Image) SetOSFeature(_ context.Context, feature ...string) {
 }
 
+// Set the entrypoint of the image.
+func (image *Image) SetEntrypoint(_ context.Context, entrypoint string) {
```
In commit (`feat: Populate application entrypoint from Kraftfile`) message `packaginf` -> `packaging`

craciunoiuc commented on action.yml:
```diff
@@ -55,8 +55,8 @@ inputs:
     required: false
-  initrd:
-    description: Include an initrd at path.
+  rootfs:
```
In commit `feat!: Introduce dynamic initramfs serialization for root filesystems` `underelying` -> `underlying`

craciunoiuc commented on initrd/dockerfile.go:
```diff
@@ -0,0 +1,202 @@
+	workdir    string
+}
+
+// NewFromDockerfile accepts an input path which represents a Dockerfile that is
```
`is` -> ` `

craciunoiuc reviewed (changes requested):
Went through the whole PR

Feels like it needs to be split up into at least 2 PRs, maybe 3

1 with the fixes, 1 with the spec update and 1 with the initrd work

nderjung committed:
feat: Implement `volume` element of the v0.6 spec update

This commit realises the adjustments to the v0.6 spec for the `volumes`
element which is intended to be used as part of both the runtime of the
unikernel.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(build): Abstract `kraft build` into interface implementations

This commit refactors the subcommand `kraft build` into an interface
`builder` which is developed in a similar vein of the interface to
`kraft run`'s `runner` such that different contexts can be used to
determine whether the provided working directory or positional
arguments are "buildable".

In this first iteration, the existing context of a working directory
populated with a `Kraftfile` that contains a `unikraft` element is
abstracted.  This is the current mode for building a Unikraft project,
however future contexts exist as well which can be developed, including:
`Kraftfile`'s that only supply a `template` element; `Kraftfile`'s that
only supply a `runtime` element (and do not construct a unikernel but do
construct a root filesystem); or the absense of a `Kraftfile` entirely
and intelligently determining the relevant template based on working
directory source files (such as determining a Go-based by project by
seeking a top-level `go.mod`).

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc commented on initrd/dockerfile.go:
```diff
@@ -0,0 +1,224 @@
+			Mode:     cpio.FileMode(d.Type().Perm()),
+			Size:     info.Size(),
+		}); err != nil {
+			return fmt.Errorf("could not writer CPIO header: %w", err)
```
`write`

craciunoiuc commented on initrd/dockerfile.go:
```diff
@@ -0,0 +1,224 @@
+			return fmt.Errorf("could not read file: %w", err)
+		}
+
+		if err := writer.WriteHeader(&cpio.Header{
+			Name:     internal,
+			Linkname: targetLink,
+			Mode:     cpio.FileMode(d.Type().Perm()),
+			Size:     info.Size(),
```
Are permissions in initramfs fixed now?

nderjung replied:
Maybe 🤷

nderjung committed:
feat: Introduce rootfs via `Dockerfile` and BuildKit

This commit introduces the ability to serialize a rootfs into a CPIO
archive by providing a `Dockerfile` as source.  This is enabled by
connecting to a BuildKit[0] instance either via UNIX socket or via a TCP
connection.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(pkg): Fix example with non-existent flag

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(pkg): Migrate `workdir` to internal command options

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(pkg): Rework packaging into characterizable packagers

This commit reworks the `kraft pkg` subcommand to accept different
packaging implementations.  By supplying an interface with the
pre-determination method `Packagable`, a candidate packaging mode can be
selected and run.  This paves the way for packaging inputs based on
different contexts.

For now identical functionality is of the `kraft pkg` remains.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(pkg): Abstract away setting up the project based on options

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(pkg): Migrate package manager and pack options to command opts

This allows their instantiation to be used by other packager providers.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(pkg): Add dynamic rootfs generation to `kraft pkg`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(pkg): Use the target's arguments if `--args` is unset

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Incorporate the name of the manifest in the platform checksum

Images can be built with the same configuration but be labelled
differently.  Account for this by including the manifest reference in
the `PlatformChecksum`.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Handle case where tag is set with `WithVersion`

This commit fixes an issue where the OCI package manager's `Catalog`
method is called with the OCI image name's tag in the version attribute
of the query.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(oci): Set the source in the trace log when adding layers

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Set the OS features after parsing the manifest

In cases where the manifest is discovered via parsing of an index, the
OSFeatures would be set in the embedded manifest structure after the
logging of the set kconfig options.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
docs(oci): Set the source and destination in the debug log for initrd

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Parse the manifest's OS features list for KConfig options

Using `NewPackageFromOCIManifestDigest` would otherwise then return
`nil` on the `KConfig` method.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
docs(oci): Set the source when logging layering the kernel

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Set the full image reference in `String`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
docs(oci): Add more verbose error message

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(pkg): Introduce `runtime` packager

This commit introduces a new packager which recognises the `runtime`
attribute of a `Kraftfile`.  This allows users to re-package images
based on an existing pre-built unikernel.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(pkg): Only prompt when list of targets is greater than 1

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc commented on unikraft/target/transform.go:
```diff
@@ -17,15 +18,31 @@ import (
-		uk := unikraft.FromContext(ctx)
-		t := TargetConfig{}
+	case string:
+		split := strings.SplitN(value, "/", 2)
+
```
Can cause sigsegv it's not tested

craciunoiuc commented on oci/pack.go:
```diff
@@ -171,7 +171,7 @@ func NewPackageFromTarget(ctx context.Context, targ target.Target, opts ...packm
 		)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("could build layer from file: %w", err)
```
`not`

nderjung committed:
fix(github-action): Resort to `oci` after `target.Format` deprecation

The target's attribute `Format` was deprecated in #873.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat: Output warning when spec version is not latest

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix: Properly handle `template` element in `Kraftfile`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(manifest): Correctly handle `--git` flag and use in `kraft build`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(component): Better parsing of string properties

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
docs(build): Add better error message regarding no targets

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc reviewed (approved):
All good here. Thanks!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
Approved-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
